producer/writer/retryable_connection.go

```diff
@@ -0,0 +1,259 @@
+)
+
+type retryableConnectionMetrics struct {
+	resetConn         tally.Counter
+	resetError        tally.Counter
+	connectError      tally.Counter
+	setKeepAliveError tally.Counter
+}
+
+func newRetryableConnectionMetrics(scope tally.Scope) retryableConnectionMetrics {
+	return retryableConnectionMetrics{
+		resetConn:         scope.Counter("reset-conn"),
+		resetError:        scope.Counter("reset-conn-error"),
+		connectError:      scope.Counter("connect-error"),
+		setKeepAliveError: scope.Counter("set-keep-alive-error"),
+	}
+}
+
+type connectFn func(addr string) (net.Conn, error)
+
+type retryableConnection struct {
+	addr            string
+	retrier         retry.Retrier
+	opts            Options
+	resetDelayNanos int64
+	logger          log.Logger
+
+	connLock       sync.RWMutex
+	conn           net.Conn
+	w              *bufio.Writer
+	r              *bufio.Reader
+	lastResetNanos int64
+	initialized    bool
+	closed         *atomic.Bool
+	doneCh         chan struct{}
+	resetCh        chan struct{}
+	wg             sync.WaitGroup
+	m              retryableConnectionMetrics
+
+	nowFn     clock.NowFn
+	connectFn connectFn
+}
+
+// newRetryableConnection creates a connection that is automatically
+// reconnected on errors.
+func newRetryableConnection(
+	addr string,
+	opts Options,
+) *retryableConnection {
+	if opts == nil {
+		opts = NewOptions()
+	}
+	c := &retryableConnection{
+		addr:            addr,
+		retrier:         retry.NewRetrier(opts.ConnectionRetryOptions().SetForever(defaultRetryForever)),
+		opts:            opts,
+		resetDelayNanos: int64(opts.ConnectionResetDelay()),
+		logger:          opts.InstrumentOptions().Logger(),
+		w:               bufio.NewWriterSize(nil, opts.ConnectionWriteBufferSize()),
+		r:               bufio.NewReaderSize(nil, opts.ConnectionReadBufferSize()),
+		lastResetNanos:  0,
+		initialized:     false,
+		closed:          atomic.NewBool(false),
+		doneCh:          make(chan struct{}),
+		resetCh:         make(chan struct{}, 1),
+		m:               newRetryableConnectionMetrics(opts.InstrumentOptions().MetricsScope()),
+		nowFn:           time.Now,
+	}
+
+	c.connectFn = c.connectOnce
+	if err := c.tryResetWithConnectFn(c.connectFn); err != nil {
+		c.SignalReset()
+	}
+	return c
+}
+
+func (c *retryableConnection) Read(p []byte) (int, error) {
+	c.connLock.RLock()
+	if !c.initialized {
+		c.connLock.RUnlock()
+		return 0, errConnNotInitialized
+	}
+	n, err := c.r.Read(p)
+	c.connLock.RUnlock()
+	return n, err
+}
+
+func (c *retryableConnection) Write(p []byte) (int, error) {
+	c.connLock.RLock()
+	if !c.initialized {
+		c.connLock.RUnlock()
+		return 0, errConnNotInitialized
+	}
+	n, err := c.w.Write(p)
+	c.connLock.RUnlock()
+	return n, err
+}
+
+func (c *retryableConnection) SignalReset() {
+	select {
+	case c.resetCh <- struct{}{}:
+	default:
+	}
+}
+
+func (c *retryableConnection) Init() {
+	c.wg.Add(1)
+	go func() {
+		c.resetConnectionForever()
+		c.wg.Done()
```

Comment:
nit: move `c.wg.Done()` to `resetConnectionForever` and just call `go c.resetConnectionForever` here.

Reply:
Same